FerramAerospaceResearch/FARGUI/FARFlightGUI/PhysicsCalcs.cs

```diff
@@ -215,17 +215,43 @@ private void CalculateVesselOrientation(Vector3d velVectorNorm)
             if (double.IsNaN(vesselInfo.sideslipAngle))
                 vesselInfo.sideslipAngle = 0;
 
-            if (_navball == null)
-                GetNavball();
-            if (!_navball)
-                return;
-            Quaternion vesselRot = Quaternion.Inverse(_navball.relativeGymbal);
+            // if (_navball == null)
+            //     GetNavball();
+            // if (!_navball)
+            //     return;
+            // Quaternion vesselRot = Quaternion.Inverse(_navball.relativeGymbal);
+
+            // vesselInfo.headingAngle = vesselRot.eulerAngles.y;
+            // vesselInfo.pitchAngle =
+            //     vesselRot.eulerAngles.x > 180 ? 360 - vesselRot.eulerAngles.x : -vesselRot.eulerAngles.x;
+            // vesselInfo.rollAngle =
+            //     vesselRot.eulerAngles.z > 180 ? 360 - vesselRot.eulerAngles.z : -vesselRot.eulerAngles.z;
+
+            // Gives the same numbers as the above, but also works in IVA mode which doesn't have a NavBall instance.
+            var localUp = (_vessel.transform.position - _vessel.mainBody.transform.position).normalized;
+            var east = Vector3.Cross(localUp, _vessel.mainBody.RotationAxis).normalized;
+            var north = Vector3.Cross(east, localUp).normalized;
+            vesselInfo.headingAngle = (SignedAngle(north, Vector3.ProjectOnPlane(up, localUp).normalized, localUp) + 360f) % 360f;
+            vesselInfo.pitchAngle = 90f - SignedAngle(localUp, up, Vector3.Cross(localUp, up).normalized);
+            vesselInfo.rollAngle = SignedAngle(Vector3.Cross(localUp, up).normalized, right, -up);
```

Comment:
Why not use `Quaternion` methods here?

Reply:
I couldn't find the correct quaternion to use and you're immediately converting them to Euler angles anyway.

Reply:
Maybe something like this
```csharp
var offsetAngles = new Vector3(_vessel.isEVA ? 90f : 0f, 0f, 0f);
var offset = Quaternion.Euler(offsetAngles);
Quaternion attitude = offset * Quaternion.Inverse(refTransform.rotation);
Transform transform = _vessel.mainBody.transform;
Vector3 position = transform.position;
Vector3 refPosition = refTransform.position;
Vector3 localUp = (refPosition - position).normalized;
var vesselRot = Quaternion.Inverse(attitude *
                                   Quaternion.LookRotation(Vector3.ProjectOnPlane(position +
                                                                          (Vector3d)transform.up *
                                                                          _vessel.mainBody.Radius -
                                                                          refPosition,
                                                                      localUp)
                                                                  .normalized,
                                                           localUp));
```

Reply:
Please do.

Reply:
It looks like the quaternion you give above doesn't give the correct angles. Here's the FAR flight data GUI extended with the results of the orientation calculations at the bottom and the KER vessel GUI for confirmation of the orientation (KER seems to be a bit inaccurate). The `err` values are the difference between the original code with `NavBall` and the two approaches above.
![Screenshot_20220502_193121](https://user-images.githubusercontent.com/18440126/166296031-808d628d-6bc2-45e3-8fe5-0e29996636df.png)